### PR TITLE
NMS-13574: External Authentication Documentation

### PR DESCRIPTION
**Pull request by mmahacek**
### All Contributors

* [X] Have you read our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [X] Have you (electronically) signed the [OpenNMS Contributor Agreement](https://cla-assistant.io/OpenNMS/opennms)?

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-13574

### CODE REVIEW

**Comment by mmahacek** [comment]
Note: needs https://github.com/OpenNMS/opennms/pull/3870 merged first to clear the antora build

File: docs/modules/operation/pages/admin/external-auth.adoc [hunk 1]
```diff
@@ -0,0 +1,88 @@
+[[ga-admin-external-auth]]
+= External Authentication
+
+If you have an existing directory for authenticating users, {page-component-title} can be configured to utilize that system for granting users access to login, in addition to allowing login from users defined directly within your {page-component-title} server.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/admin/external-auth.adoc:4]:
```suggestion
If you have an existing directory for authenticating users, {page-component-title} can be configured to use that system to grant users access to log in, in addition to allowing log in from users defined directly within your {page-component-title} server.
```

File: docs/modules/operation/pages/admin/external-auth.adoc [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+Skeleton files to use as a starter for your configuration.
+The "disabled" files are ignored when Spring Security loads.
+
+== Anatomy of an LDAP configuration
+
+Whether the authentication source is an Active Directory, a Novell eDirectory, or some other LDAP-enabled directory, the basic components (expressed as "beans" in Spring Framework lingo) that need to exist and be configured are the same.
+
+While Active Directory and OpenLDAP (and others) implement the same LDAP protocol, their schemas are substantially different.
+The biggest difference is the `memberOf` attribute, which is not supported by default LDAP v3 specification, and is not present in common OpenLDAP installs, but comes standard with AD.
+This has a significant implication in that you cannot filter a user search by nested group membership, and most installs tend not to be fully hierarchical (User's DN doesn't cite a home group).
```
Comment (Bonrob2) [inline docs/modules/operation/pages/admin/external-auth.adoc:55]:
```suggestion
This has a significant implication in that you cannot filter a user search by nested group membership, and most installs tend not to be fully hierarchical (user's DN doesn't cite a home group).
```

File: docs/modules/operation/pages/admin/external-auth.adoc [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+Provide credentials to bind to your directory in the `defaultUser` and `defaultPassword` properties.
+
+userSearch::
+Tells Spring Security where to find the users in your directory.
+
+userGroupLdapAuthoritiesPopulator::
+Tells Spring Security where to find groups in your directory.
+By defining `groupToRoleMap` entries, groups in your directory can be mapped to one or more security roles to determine what level of access the user should be granted.
+
+NOTE: If your directory requires a SSL connection, and is signed with a private certificate authority, make sure to add the certificates to your xref:admin/https/https-client.adoc[truststore].
```
Comment (Bonrob2) [inline docs/modules/operation/pages/admin/external-auth.adoc:72]:
```suggestion
NOTE: If your directory requires an SSL connection, and is signed with a private certificate authority, make sure to add the certificates to your xref:admin/https/https-client.adoc[truststore].
```

File: docs/modules/operation/pages/admin/external-auth.adoc [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+After making any changes to Spring Security configuration, {page-component-title} must be restarted for the changes to take effect.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/admin/external-auth.adoc:74]:
```suggestion
After making any changes to Spring Security configuration, you must restart {page-component-title} for the changes to take effect.
```

File: docs/modules/operation/pages/admin/external-auth.adoc [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+Before doing this, it's a good idea to check that your changes have not led to malformed XML files:
+
+[source, console]
+----
+xmllint --noout applicationContext-spring-security.xml spring-security.d/*.xml
+----
+
+If this command produces no output, then the XML files are well-formed.
+If you've gotten all the details right, you'll be able to log in to the webapp with your directory credentials after restarting OpenNMS.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/admin/external-auth.adoc:83]:
```suggestion
If you have all the details right, you'll be able to log in to the webapp with your directory credentials after restarting OpenNMS.
```

**Review by Bonrob2 — CHANGES_REQUESTED** [review]
Minor changes.

**Review by ProjectJYL — DISMISSED** [review]
LGTM

File: docs/modules/operation/pages/admin/external-auth.adoc [hunk 6]
```diff
@@ -0,0 +1,88 @@
+[[ga-admin-external-auth]]
+= External Authentication
+
+If you have an existing directory for authenticating users, {page-component-title} can be configured to use that system to grant users access to log in, in addition to allowing log in from users defined directly within your {page-component-title} server.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/admin/external-auth.adoc:4]:
```suggestion
If you have an existing directory for authenticating users, you can configure {page-component-title} to use that system to grant users access to log in, in addition to allowing log in from users defined directly within your {page-component-title} server.
```

File: docs/modules/operation/pages/admin/external-auth.adoc [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+== Configuring external authentication
+
+The {page-component-title} webapp uses link:https://spring.io/projects/spring-security[Spring Security] for its user authentication, authorization, and accounting (AAA) concerns.
+Configuring external authentication requires configuring Spring Security correctly for your own environment.
+
+The Spring Security configuration can be modified without making large, invasive changes to the main context file.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/admin/external-auth.adoc:11]:
```suggestion
You can modify the Spring Security configuration without making large, invasive changes to the main context file.
```

File: docs/modules/operation/pages/admin/external-auth.adoc [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+authenticationSource::
+Provide credentials to bind to your directory in the `defaultUser` and `defaultPassword` properties.
+
+userSearch::
+Tells Spring Security where to find the users in your directory.
+
+userGroupLdapAuthoritiesPopulator::
+Tells Spring Security where to find groups in your directory.
+By defining `groupToRoleMap` entries, groups in your directory can be mapped to one or more security roles to determine what level of access the user should be granted.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/admin/external-auth.adoc:70]:
```suggestion
By defining `groupToRoleMap` entries, groups in your directory can be mapped to one or more security roles to determine what level of access to grant to the user.
```

**Review by Bonrob2 — CHANGES_REQUESTED** [review]
Minor changes. It looks great.

**Review by Bonrob2 — APPROVED** [review]
LGTM!

**Pull request merged**